### PR TITLE
Progress bar for downloading files

### LINKED ISSUE #466: need some indicator to highlight the S2Lib installation process
as it takes quite long before anything actually starts to happen on the UI (at least in the case of S2Lib which is about 30MB)

### PR DESCRIPTION
**Pull request by jaakkonarhi**
# Description of the PR

+ Shows the name of the file being downloaded in the bar
+ Closes #466 

# Testing
<table>
  <tr>
    <th>unit</th>
    <th>integration</th>
    <th>e2e</th>
    <th>manual</th>
  </tr>
  <tr>
    <td>
      <ul>
        <li>- [ ] new <b>unit</b> tests created</li>
        <li>- [x] all <b>unit</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>integration</b> tests created</li>
        <li>- [x] all <b>integration</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>e2e</b> tests created</li>
        <li>- [x] all <b>e2e</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [x] <b>manual</b> testing went well</li>
      </ul>
    </td>
  </tr>
</table>  
  
# Have you updated the [TESTING.md](https://github.com/Aalto-LeTech/intellij-plugin/blob/master/TESTING.md) or other relevant documentation on _your_ branch?

- [ ] Yes.
- [ ] Not yet. I will do it next.
- [x] Not relevant.

### CODE REVIEW

**Review by Taikelenn — COMMENTED** [review]
Does this really solve the linked issue? This adds a progress bar for the zip extraction operation, not downloading. The issue is about the progress for the download operation.

Extracting a zip file rarely takes more than a few hundred milliseconds, so most users won't even notice the progress bar.

File: src/main/java/fi/aalto/cs/apluscourses/utils/RemoteFileCache.java [hunk 1]
```diff
@@ -79,14 +87,49 @@ public boolean isInvalid() {
...
+                                byte @NotNull [] buffer,
+                                long length,
+                                @NotNull URL url,
+                                @NotNull Project project)
+      throws IOException {
+    var progress = PluginSettings
+        .getInstance()
+        .getMainViewModel(project)
+        .progressViewModel
+        .start((int) length / 1024,
```
Comment (Taikelenn) [inline src/main/java/fi/aalto/cs/apluscourses/utils/RemoteFileCache.java:126]: Wrap the expression in parentheses to enable terabyte downloads

File: src/main/java/fi/aalto/cs/apluscourses/utils/RemoteFileCache.java [hunk 2]
```diff
@@ -79,14 +87,49 @@ public boolean isInvalid() {
...
+            getAndReplaceText("ui.ProgressBarView.downloading", url.getFile()), false);
+    int n;
+    while (EOF != (n = input.read(buffer))) {
+      output.write(buffer, 0, n);
+      progress.incrementBy(n / 1024);
```
Comment (Taikelenn) [inline src/main/java/fi/aalto/cs/apluscourses/utils/RemoteFileCache.java:131]: If the number of read bytes is smaller than 1024, the progress bar won't move - fix pls

File: src/main/java/fi/aalto/cs/apluscourses/utils/RemoteFileCache.java [hunk 3]
```diff
@@ -79,14 +87,49 @@ public boolean isInvalid() {
...
+                                    @NotNull File destination,
+                                    int connectionTimeout,
+                                    final int readTimeout,
+                                    @NotNull Project project) throws IOException {
+    var connection = source.openConnection();
+    connection.setConnectTimeout(connectionTimeout);
+    connection.setReadTimeout(readTimeout);
+    var length = connection.getContentLengthLong();
+    var stream = connection.getInputStream();
+    var out = openOutputStream(destination);
```
Comment (Taikelenn) [inline src/main/java/fi/aalto/cs/apluscourses/utils/RemoteFileCache.java:111]: Consider using try-with-resources

**Review by stellatsv — APPROVED** [review]
Very nice!

**Review by Taikelenn — COMMENTED** [review]
Almost good, but please handle network failures

**Pull request merged**